pkg/util/id.go

```diff
@@ -16,31 +16,83 @@ limitations under the License.
 package util
 
 import (
+	"fmt"
 	"io/ioutil"
+	"os"
+	"path"
 	"strings"
+
+	"github.com/google/uuid"
 )
 
-const maxMachineIDLength = 12
+const (
+	maxNodeIDLength = 12
+)
 
-func GetMachineID() (string, error) {
-	buf, err := ioutil.ReadFile("/etc/machine-id")
+// To ensure a constant node ID between runs, the id will be persisted in the data directory.
+func LoadPersistedNodeID(dataDir string) (string, error) {
+	id, save, err := loadNodeID(dataDir)
 	if err != nil {
-		return "", err
+		return "", fmt.Errorf("failed to load node id. %+v", err)
+	}
+
+	if save {
+		err = saveNodeID(dataDir, id)
+		if err != nil {
+			return "", fmt.Errorf("failed to save node id. %+v", err)
+		}
+	}
+
+	return id, nil
+}
+
+// Attempt to load the node from the data dir. If not found,
+// generate a new random id
+func loadNodeID(dataDir string) (string, bool, error) {
+	// first read the id from the cache
+	buf, err := ioutil.ReadFile(path.Join(dataDir, "id"))
+	if err == nil {
+		// successfully loaded the id from the file
+		id := strings.TrimSpace(string(buf))
+		if len(id) == 0 {
+			return "", false, fmt.Errorf("the cached node id cannot be empty")
```

Comment:
hopefully the user will never see this error because it means the id file was corrupted, but i'll clarify